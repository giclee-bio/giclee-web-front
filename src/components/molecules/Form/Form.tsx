'use client'

import * as RadixForm from '@radix-ui/react-form'
import { HTMLInputTypeAttribute } from 'react'

import Button from '@/components/atoms/Button'

import type { Props as ButtonProps } from '@/components/atoms/Button'

export type FormField = {
  label: string
  name: string
  require?: boolean
  tag: 'input' | 'textarea'
  type: HTMLInputTypeAttribute
  validations?: FormMessage[]
} & ({ require: undefined; tag: 'textarea' } | { tag: 'input' })

export type FormMessage = {
  match: RadixForm.CustomMatcher
  message: string
}

type Props = {
  buttonProps: ButtonProps
  formFields: FormField[]
}

const FormMessage: React.FC<FormMessage> = ({ match, message }) => {
  return (
    <RadixForm.Message className='text-ruby-400' match={match}>
      {message}
    </RadixForm.Message>
  )
}

const FormItem: React.FC<FormField> = ({
  label,
  name,
  require = false,
  tag,
  type,
  validations,
}) => {
  return (
    <RadixForm.Field name={name}>
      <div className='flex justify-between text-sm'>
        <RadixForm.Label className='font-bold'>{label}</RadixForm.Label>
        {require && (
          <RadixForm.Message className='text-ruby-400' match='valueMissing'>
            必須項目です
          </RadixForm.Message>
        )}
        {validations &&
          validations.map((validation, key) => {
            return <FormMessage {...validation} key={key} />
          })}
      </div>
      <RadixForm.Control asChild className='mt-2 bg-white'>
        {tag === 'input' ? (
          <input className='h-8 w-full rounded px-3 text-base' required={require} type={type} />
        ) : (
          <textarea className='h-8 w-full rounded px-3 text-base' required={require} />
        )}
      </RadixForm.Control>
    </RadixForm.Field>
  )
}

const Form: React.FC<Props> = ({ buttonProps, formFields }) => {
  return (
    <RadixForm.Root className='w-64'>
      {formFields.map((formFieid, key) => {
        return <FormItem {...formFieid} key={key} />
      })}
      <RadixForm.Submit asChild>
        <Button {...buttonProps} className='mx-auto mt-9 block' />
      </RadixForm.Submit>
    </RadixForm.Root>
  )
}

export default Form
