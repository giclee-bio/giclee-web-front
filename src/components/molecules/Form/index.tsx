import Form from './Form'

import type { FormField } from './Form'

export default Form
export type { FormField }
