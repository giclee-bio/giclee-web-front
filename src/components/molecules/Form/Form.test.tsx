import { render } from '@testing-library/react'

import Form from './Form'

import type { Props as ButtonProps } from '@/components/atoms/Button'
import type { FormField } from '@/components/molecules/Form'

const buttonProps: ButtonProps = {
  children: '登録する',
  type: 'primary',
}

const formFields: FormField[] = [
  {
    label: 'ユーザー名',
    name: 'username',
    require: true,
    tag: 'input',
    type: 'text',
  },
]

describe('Form', () => {
  it('snapshot', async () => {
    const { asFragment } = render(<Form buttonProps={buttonProps} formFields={formFields} />, {})
    expect(asFragment()).toMatchSnapshot()
  })
})
