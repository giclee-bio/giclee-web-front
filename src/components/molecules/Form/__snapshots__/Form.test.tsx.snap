// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshot 1`] = `
<DocumentFragment>
  <form
    class="w-64"
  >
    <div>
      <div
        class="flex justify-between text-sm"
      >
        <label
          class="font-bold"
          for="radix-:r0:"
        >
          ユーザー名
        </label>
      </div>
      <input
        class="mt-2 bg-white h-8 w-full rounded px-3 text-base"
        id="radix-:r0:"
        name="username"
        required=""
        title=""
        type="text"
      />
    </div>
    <button
      class="mx-auto mt-9 block h-10 rounded text-center font-bold leading-10 w-64 bg-primary text-base"
    >
      登録する
    </button>
  </form>
</DocumentFragment>
`;
